### PR TITLE
Add nether survival games

### PR DESCRIPTION
**Pull request by yitzy299**
Has nether wastes, crimson and warped forest biomes, basalt deltas and soul sand valleys

- Piglins and gold ore for another interesting gameplay element
- New optional config value for generation of nether structures, defaults to 0
- New optional config value for dimension of map, defaults to overworld

### CODE REVIEW

File: src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java [hunk 1]
```diff
@@ -58,268 +58,290 @@
 public class SurvivalGamesChunkGenerator extends GameChunkGenerator {
 
 
-	private final NoiseSampler2d treeDensityNoise;
+    private final NoiseSampler2d treeDensityNoise;
```
Comment (Gegy) [inline src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java:61]: Oh no

Comment (yitzy299) [reply src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java:61]: That is very concerning, bad GitHub

File: src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java [hunk 2]
```diff
@@ -58,268 +58,290 @@
...
+
+            BlockPos start = new BlockPos(startX, 0, startZ);
+            ChunkPos chunkPos = new ChunkPos(start);
+
+            if (mask.isIn(chunkPos)) {
+                continue;
+            }
+
+            SurvivalGamesJigsawGenerator outskirtGenerator = new SurvivalGamesJigsawGenerator(server, this, piecesByChunk);
+            outskirtGenerator.arrangePieces(start, new Identifier("survivalgames", "nether_outskirts"), 0);
```
Comment (SuperCoder7979) [inline src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java:142]: This should go into the config, so you can specify what type of outskirts you want

File: src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java [hunk 3]
```diff
@@ -58,268 +58,290 @@
...
+                            height = (int) MathHelper.clampedLerp(height, piece.getPos().getY(), 1 - rad);
+                        }
+                    }
+                }
+
+                // Generation height ensures that the generator iterates up to at least the water level.
+                int genHeight = Math.max(height, 48);
+                for (int y = 0; y <= genHeight; y++) {
+                    // Simple surface building
+                    BlockState state = Blocks.STONE.getDefaultState();
```
Comment (SuperCoder7979) [inline src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java:208]: might need to be put into the config for it to be netherrack

File: src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java [hunk 4]
```diff
@@ -58,268 +58,290 @@
...
+                        }
+                    } else if ((height - y) <= 3) { //TODO: biome controls under depth
+                        state = biome.underState(random, x, z);
+                    } else if (y == 0) {
+                        state = Blocks.BEDROCK.getDefaultState();
+                    }
+
+                    // If the y is higher than the land height, then we must place water
+                    if (y > height) {
+                        state = Blocks.WATER.getDefaultState();
```
Comment (SuperCoder7979) [inline src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java:225]: Putting this into the config would be good too

File: src/main/resources/data/survivalgames/games/nether.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    "safe_secs": 120,
+    "shrink_secs": 240
+  },
+  "players": {
+    "min": 1,
+    "max": 100
+  },
+  "town_depth": -6,
+  "outskirts_building_count": 0,
+  "nether_building_count": 10,
```
Comment (SuperCoder7979) [inline src/main/resources/data/survivalgames/games/nether.json:17]: Should be generified into "outskirts_building_count" and "outskirts_pool" as stated above

**Review by SuperCoder7979 — CHANGES_REQUESTED** [review]

File: src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java [hunk 6]
```diff
@@ -58,268 +58,278 @@
...
+
+            BlockPos start = new BlockPos(startX, 0, startZ);
+            ChunkPos chunkPos = new ChunkPos(start);
+
+            if (mask.isIn(chunkPos)) {
+                continue;
+            }
+
+            SurvivalGamesJigsawGenerator outskirtGenerator = new SurvivalGamesJigsawGenerator(server, this, piecesByChunk);
+            outskirtGenerator.arrangePieces(start, new Identifier("survivalgames", config.dimension.getPath() + "_outskirts_buildings"), 0);
```
Comment (SuperCoder7979) [inline src/main/java/supercoder79/survivalgames/game/map/SurvivalGamesChunkGenerator.java:130]: Ideally we would have a custom codec entry for the outskirts path. That'll allow us to eventually add tropical buildings to tropical maps for example

**Review by SuperCoder7979 — CHANGES_REQUESTED** [review]

**Review by SuperCoder7979 — APPROVED** [review]
Still needs a few fixes, but will do those myself. Thanks for this!

**Comment by yitzy299** [comment]
Thanks!

**Pull request merged**